hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/transaction/IngestionPrimaryWriterBasedConflictResolutionStrategy.java

```diff
@@ -0,0 +1,116 @@
+   */
+  @Override
+  public Stream<HoodieInstant> getCandidateInstants(HoodieTableMetaClient metaClient, HoodieInstant currentInstant,
+                                                    Option<HoodieInstant> lastSuccessfulInstant) {
+    HoodieActiveTimeline activeTimeline = metaClient.reloadActiveTimeline();
+    if ((REPLACE_COMMIT_ACTION.equals(currentInstant.getAction())
+          && ClusteringUtils.isClusteringCommit(metaClient, currentInstant))
+        || COMPACTION_ACTION.equals(currentInstant.getAction())) {
+      return getCandidateInstantsForTableServicesCommits(activeTimeline, currentInstant);
+    } else {
+      return getCandidateInstantsForNonTableServicesCommits(activeTimeline, currentInstant);
+    }
+  }
+
+  private Stream<HoodieInstant> getCandidateInstantsForNonTableServicesCommits(HoodieActiveTimeline activeTimeline, HoodieInstant currentInstant) {
+
+    // To findout which instants are conflicting, we apply the following logic
+    // Get all the completed instants timeline only for commits that have happened
+    // since the last successful write based on the transition times.
+    // We need to check for write conflicts since they may have mutated the same files
+    // that are being newly created by the current write.
+    List<HoodieInstant> completedCommitsInstants = activeTimeline
+        .getTimelineOfActions(CollectionUtils.createSet(COMMIT_ACTION, REPLACE_COMMIT_ACTION, COMPACTION_ACTION, DELTA_COMMIT_ACTION))
+        .filterCompletedInstants()
+        .findInstantsModifiedAfterByStateTransitionTime(currentInstant.getTimestamp())
+        .getInstantsOrderedByStateTransitionTime()
+        .collect(Collectors.toList());
+    LOG.info(String.format("Instants that may have conflict with %s are %s", currentInstant, completedCommitsInstants));
+    return completedCommitsInstants.stream();
+  }
+
+  /**
+   * To find which instants are conflicting, we apply the following logic
+   * Get both completed instants and ingestion inflight commits that have happened since the last successful write.
+   * We need to check for write conflicts since they may have mutated the same files
+   * that are being newly created by the current write.
+   */
+  private Stream<HoodieInstant> getCandidateInstantsForTableServicesCommits(HoodieActiveTimeline activeTimeline, HoodieInstant currentInstant) {
+    // Fetch list of completed commits.
+    Stream<HoodieInstant> completedCommitsStream =
+        activeTimeline
+            .getTimelineOfActions(CollectionUtils.createSet(COMMIT_ACTION, REPLACE_COMMIT_ACTION, COMPACTION_ACTION, DELTA_COMMIT_ACTION))
+            .filterCompletedInstants()
+            .findInstantsModifiedAfterByStateTransitionTime(currentInstant.getTimestamp())
+            .getInstantsAsStream();
+
+    // Fetch list of ingestion inflight commits.
+    Stream<HoodieInstant> inflightIngestionCommitsStream =
```

Comment:
@danny0405 @suryaprasanna 
One question, Whether the ingestion commit is inflight or request, there is no useful information in its metadata, how to use it for conflict detection?  If this prerequisite does not exist, then the `IngestionPrimaryWriterBasedConflictResolutionStrategy` will be invalid